### PR TITLE
#173 refactored file producer for s3 producer reusage [PART 1]

### PR DESCRIPTION
**Pull request by msiebeneicher**
Here the first PR for task #173 

This PR mostly contain the refactoring of the file producer.

### CODE REVIEW

**Comment by msiebeneicher** [comment]
There is now an issue in the "AggregatePipeline". I will investigate the problem next week - i guess that is related with some configs

File: core/components/batchedWriterAssembly.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	writer          BatchedWriter
+	assembly        core.WriterAssembly
+	flushTimeout    time.Duration // max sec to wait before a flush is aborted
+	batchTimeout    time.Duration // max sec to wait before batch will flushed
+	batchFlushCount int
+	logger          logrus.FieldLogger
+}
+
+// BatchedWriter is an interface for different file writer like disk, s3, etc.
+// TODO: Replace methods with os.FileInfo interface
```
Comment (arnecls) [inline core/components/batchedWriterAssembly.go]: I think this interface is fine as it is.
os.FileInfo introduces functions not required by the functions using BatchedWriter.
Function that are not used should not be part if the interface.

File: core/components/batchedWriterAssembly.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// SetWriter set a BatchedWriter interface implementation
+func (bwa *BatchedWriterAssembly) SetWriter(writer BatchedWriter) {
+	bwa.writer = writer
+	bwa.Created = time.Now()
+}
+
+// UnsetWriter unset the current writer
+func (bwa *BatchedWriterAssembly) UnsetWriter() {
```
Comment (arnecls) [inline core/components/batchedWriterAssembly.go]: Does this function have to be public? Do we even require it?

File: producer/file.go [hunk 3]
```diff
@@ -165,209 +140,222 @@ func (prod *File) Configure(conf core.PluginConfigReader) {
...
+	for _, batchedFile := range prod.files {
+		batchedFile.FlushOnTimeOut()
 	}
 }
 
 func (prod *File) writeMessage(msg *core.Message) {
-	streamMsg := msg.Clone()
-
-	state, err := prod.getFileState(streamMsg.GetStreamID(), false)
+	batchedFile, err := prod.getBatchedFile(msg.GetOrigStreamID(), false)
```
Comment (arnecls) [inline producer/file.go]: This should be GetStreamID as the "last" stream should define the file name

File: core/components/batchedWriterAssembly.go (hunk 2 above)
Comment (msiebeneicher) [inline core/components/batchedWriterAssembly.go]: This method was in use before i implemented the `GetWriterAndUnset()`. So currently i can change they to private but this is a bit "inconsistent" if you only provide a `GetWriter()` and `GetWriterAndUnset()` method. In my opinion the interface is now cleaner.

File: producer/file.go (hunk 3 above)
Comment (msiebeneicher) [inline producer/file.go]: right - i will change that

**Pull request merged**